### PR TITLE
[AD-242] Update state with new blocks 

### PR DESCRIPTION
**Pull request by hithroc**

### CODE REVIEW

File: ariadne/ariadne.cabal [hunk 1]
```diff
@@ -98,6 +98,7 @@ library
     , word-wrap
     , process
     , template-haskell
+    , unliftio-core
```
Comment (neongreen) [inline ariadne/ariadne.cabal]: :+1:

File: ariadne/src/Ariadne/Cardano/Face.hs [hunk 2]
```diff
@@ -63,9 +107,148 @@ data CardanoEvent
...
+instance
+  ( Mockable d Production
+  , MFunctor' d CardanoMode CardanoModeMonad
+  , MFunctor' d CardanoModeMonad Production
+  )
+  => Mockable d CardanoMode where
+  liftMockable = CardanoMode . liftMockable . hoist' unwrapCardanoMode
+
+-- ♫ E-E-G-G-E-E-G Mr. Boilerplate!
+-- Amazing, with Mr. Boilerplate my code is now twice as long!!!
```
[the remark this answers is not in the document]
Comment (hithroc) [reply ariadne/src/Ariadne/Cardano/Face.hs]: Oh, wow, I forgot I left it there!

[the remark this answers is not in the document]
Comment (Pastafarianist) [reply ariadne/src/Ariadne/Cardano/Face.hs]: Also, you may want to [brush up on Haddock syntax](https://github.com/ghc/haddock/blob/master/doc/cheatsheet/haddocks.pdf).

File: ariadne/src/Ariadne/Wallet/Backend/Mode.hs [hunk 3]
```diff
@@ -22,10 +22,10 @@ import Pos.Core.Configuration (HasConfiguration)
 import Pos.Launcher.Configuration (HasConfigurations)
 import Pos.Txp.DB.Utxo (getFilteredUtxo)
 
-import Ariadne.Cardano.Face (CardanoMode)
+import Ariadne.Cardano.Face (CardanoMode(..))
```
Comment (gromakovsky) [inline ariadne/src/Ariadne/Wallet/Backend/Mode.hs]: :thinking:

File: ariadne/src/Ariadne/Cardano/Backend.hs [hunk 4]
```diff
@@ -79,7 +84,15 @@ runCardanoNode cardanoContextVar diffusionVar commonArgs sendCardanoEvent = do
               , extractionWorker
               , statusPollingWorker sendCardanoEvent
               ]
-      runNodeReal nodeParams sscParams workers
+      let
+        realModeToCardanoMode m = CardanoMode $ ask >>= lift . runReaderT m . ccRealModeContext
+        cardanoModeToRealMode (CardanoMode m) = ask >>= lift . runReaderT m . CardanoContext bHandle
```
Comment (gromakovsky) [inline ariadne/src/Ariadne/Cardano/Backend.hs]: Can be simplified using `withReaderT`.

File: ariadne/src/Ariadne/Cardano/Face.hs [hunk 5]
```diff
@@ -56,9 +98,149 @@ data CardanoEvent
...
+    dbDelete = dbDeleteDefault
+    dbPutSerBlunds = dbPutSerBlundsRealDefault
+
+instance (HasConfiguration, HasTxpConfiguration) =>
+         MonadTxpLocal CardanoMode where
+    txpNormalize = txNormalize
+    txpProcessTx = txProcessTransaction
+
+instance MonadReporting CardanoMode where
+    report rt = Mtl.ask >>= liftIO . flip runReporter rt . (rmcReporter . ccRealModeContext)
```
Comment (gromakovsky) [inline ariadne/src/Ariadne/Cardano/Face.hs]: I think all these instances should reuse implementations for `RealMode`. In this version definitions are mostly copy-pasted from those in `Pos.WorkMode`.
See how instances are defined [here](https://github.com/input-output-hk/cardano-sl/blob/226186092edd826c469850d2d808df9720f08efc/auxx/src/Mode.hs) (using `realModeToAuxx` mostly).

File: ariadne/src/Ariadne/Wallet/Cardano/Kernel.hs [hunk 6]
```diff
@@ -135,6 +142,21 @@ initPassiveWallet logMessage db = do
...
+    mapMaybe' :: (a -> Maybe b) -> OldestFirst NE a -> OldestFirst [] b
+    mapMaybe' f = OldestFirst . mapMaybe f . toList . getOldestFirst
+    rightToJust   = either (const Nothing) Just
+    -- The use of the unsafe constructor 'UnsafeRawResolvedBlock' is justified
+    -- by the invariants established in the 'Blund'.
+    blundToResolvedBlock :: Blund -> Maybe ResolvedBlock
+    blundToResolvedBlock (b,u)
+        = rightToJust b <&> \mainBlock ->
+            fromRawResolvedBlock
+            $ UnsafeRawResolvedBlock mainBlock (map (map fromJust) $ undoTx u)
```
Comment (gromakovsky) [inline ariadne/src/Ariadne/Wallet/Cardano/Kernel.hs]: I don't like that you inlined stuff from `Cardano.Wallet.WalletLayer.Kernel` here. It increases diff with IOHK. I find it better to copy-paste `Cardano.Wallet.WalletLayer.Kernel` into our code (and some other modules probably) and use `Cardano.Wallet.WalletLayer.Kernel.bracketPassiveWallet`.

File: ariadne/src/Ariadne/Wallet/Backend.hs [hunk 7]
```diff
@@ -9,27 +9,37 @@ import Data.Acid (openLocalStateFrom, query)
...
   -- TODO: Do I need to close session on exit?
   acidDb <- openLocalStateFrom walletAcidDbPathPlaceholder defDB
-  return $ \cf@CardanoFace {..} sendWalletEvent ->
+
+  -- get the block apply handler
+  -- We're doing this dirty hack with MVar for now because we don't have
+  -- a proper bracket architecture anywhere
+  mHandle <- newEmptyMVar
+  bracketPassiveWallet (\sev -> usingLoggerName "passive-wallet" . logMessage sev) (putMVar mHandle)
+  pw <- takeMVar mHandle
```
[the remark this answers is not in the document]
Comment (hithroc) [reply ariadne/src/Ariadne/Wallet/Backend.hs]: You can't have an empty `IORef` can you? `MVar` can be empty, that's why I think it makes more sense to use it here.

[the remark this answers is not in the document]
Comment (Pastafarianist) [reply ariadne/src/Ariadne/Wallet/Backend.hs]: Well, you can have an `IORef (Maybe smth)` :wink:

Please add a comment explaining the rationale of choosing `MVar`. It is usually used in a multithreaded environment after all.

[the remark this answers is not in the document]
Comment (hithroc) [reply ariadne/src/Ariadne/Wallet/Backend.hs]: I don't see a big difference between using `MVar` or `IORef`, it's a very small implementation detail that is temporary anyway.

[the remark this answers is not in the document]
Comment (gromakovsky) [reply ariadne/src/Ariadne/Wallet/Backend.hs]: Yes, it's a temporary hack, doesn't need to be commented, I think.

File: ariadne/src/Ariadne/Cardano/Face.hs (hunk 5 above)
Comment (gromakovsky) [reply ariadne/src/Ariadne/Cardano/Face.hs]: This instance is still copy-pasted :unamused:

Comment (gromakovsky) [reply ariadne/src/Ariadne/Cardano/Face.hs]: Please fix.

File: ariadne/src/Ariadne/Cardano/Face.hs [hunk 8]
```diff
@@ -56,9 +90,153 @@ data CardanoEvent
...
+instance HasConfiguration => MonadDB CardanoMode where
+    dbPut = realModeToCardanoMode ... dbPut
+    dbWriteBatch = realModeToCardanoMode ... dbWriteBatch
+    dbDelete = realModeToCardanoMode ... dbDelete
+    dbPutSerBlunds = realModeToCardanoMode ... dbPutSerBlunds
+
+instance (HasConfiguration, HasTxpConfiguration) =>
+         MonadTxpLocal CardanoMode where
+    txpNormalize = txNormalize
+    txpProcessTx = txProcessTransaction
```
Comment (gromakovsky) [inline ariadne/src/Ariadne/Cardano/Face.hs]: Can `realModeToCardanoMode` be used in this instance as well?

File: ariadne/src/Ariadne/Wallet/Cardano/Kernel.hs [hunk 9]
```diff
@@ -76,9 +77,9 @@ import Pos.Txp (Utxo)
 --
 data PassiveWallet = PassiveWallet {
       -- | Send log message
-      _walletLogMessage :: Severity -> Text -> IO () -- ^ Logger
-    , _walletESKs       :: MVar WalletESKs           -- ^ ESKs indexed by WalletId
-    , _wallets          :: AcidState DB              -- ^ Database handle
+      _walletLogMessage :: Severity -> Text -> IO ()     -- ^ Logger
+    , _walletUS         :: TVar UserSecret -- ^ ESKs indexed by WalletId
```
Comment (gromakovsky) [inline ariadne/src/Ariadne/Wallet/Cardano/Kernel.hs]: Update comment pls.

**Pull request merged**